airflow/models/dagrun.py

```diff
@@ -218,12 +219,19 @@ def get_dag(self):
         return self.dag
 
     @provide_session
-    def get_previous_dagrun(self, session=None):
+    def get_previous_dagrun(self, session: Session = None, state: str = None) -> Optional['DagRun']:
         """The previous DagRun, if there is one"""
 
-        return session.query(DagRun).filter(
+        assert isinstance(session, Session)  # mypy
```

Comment:
Here's an example of failing mypy test output: https://travis-ci.org/apache/airflow/jobs/542113316

In this case it tries to access the `query` method on a `None`.  I think it's because `None` is the default argument for `session`.  It did not fail until I added type hinting.

Reply:
Pinging @jmcarp as our resident mypy "expert" :) Does this sound like a sensible change?

Reply:
Could you test something @dstandish: what happens if you remove the `=None` on the session and this assert? because of the decorator it will never be passed with None in anyway.

Reply:
@dstandish By default mypy treats arguments which default to `None` as implicit `Optional[T]`. IMO this is a good thing because it adds a minimal null-safety threshold to typed Python, but it creates these confusing situations.

If as @ashb asserts the decorator will _always_ pass a `Session` object I would just remove the `isinstance` call and `None` default value to the `session` argument.

As a side note, there are situations where you positively need an argument with defaults to `None`, but where you know for a fact that in practice this will never happen (i.e. some usages of `functools.partial`) In these cases you could use a mypy `cast` here instead of `isinstance`. The difference being that `cast` is a dud at runtime.

Reply:
God to know @pgagnon

Reply:
ok so @ashb shall i just change it to `cast` or shall I remove default argument of `None`? if i remove `None', i will not be able to leave `session` the last argument.

Reply:
Good point. Use `cast` then.

Reply:
OK so it turns out "using `cast`" means, for example, adding this line instead of the assert: `session = cast(Session, session)`.  Also pretty ugly!   Alas, it seems like what we have to do.

Reply:
@ashb I think it might be preferable to remove the `None` argument instead of using `cast`, because it makes the function's signature more accurate.

Reply:
@pgagnon Yes, but then it can't be after other optional arguments, and the entire rest of the codebase where we use `@provide_session` has `session=None` as the last kwarg in functions.

Reply:
@ashb Ah I see. That trumps my concern then.

Reply:
I think I'd rather we kept with convention on this PR and changed everything wholesale if we want to make a change.

Reply:
Did we decide to type everything everywhere?  I am perfectly happy to skip typing in this PR if it is raising problems we're not keen to deal with now, if we want to later fix in a universal way.
Of course, I can see a benefit to adding typing whenever you touch something, if the decision is to add typing everywhere.

Reply:
Add typing as we edit bits, but don't change the pattern we use for `session=None`